packages/nimble-components/src/select/specs/filterable-select-hld.md

```diff
@@ -0,0 +1,114 @@
+# Filterable Select
+
+## Problem Statement
+
+Our clients have a need of a filterable dropdown component that does not allow arbitrary text as a value, but only values available in the dropdown. The filtering will have the following behaviors (in the initial feature set):
+
+-   The filter text will match _any_ text within the `textContent` of each `ListOption` in a case insensitive way. Specifically, the filter text will match wholly against the target (e.g. filter text of "ad" will match "Add", but not "abcd").
+-   Each time the dropdown is opened the filter text is cleared.
+-   While the dropdown is opened all keystrokes except `<ArrowUp>`. `<ArrowDown>`, `<Enter>`, `<Home>` and `<End>` will apply to the filter.
+-   Pressing `<Esc>` will close the dropdown and revert the value to what it was prior to opening dropdown.
+
+## Links To Relevant Work Items and Reference Material
+
+-   [#1273: Filterable NimbleSelect](https://github.com/ni/nimble/issues/1273)
+-   [Filterable Select IxD](./IxD.md#filter)
+
+## Implementation / Design
+
+### API
+
+#### Select
+
+We will provide a means for clients to enable this feature with a new attribute:
+
+```ts
+export class Select() {
+    ...
+    @attr({ attribute: 'filter-mode' })
+    public filterMode = FilterMode.none;
+    ...
+}
+
+// Provide enum for filterMode to allow for future modes
+export const FilterMode  = {
+    none: 'none';
+    standard: 'standard';
+} as const;
+```
+
+-   The `standard` filterMode will result in case-insensitive, diacritic-insensitive filtering.
+-   `filterMode` will default to `none` so as not to affect existing clients.
```

Comment:
Does a filterMode of none mean that the filterbox is hidden?
Guess I'm wondering if we should make none the default.. I think by default it should be visible. Do we have usage guidance for when it is hidden? If we made it a breaking change to show by default do we know the risks?

Reply:
Good point, I hadn't thought about usage guidance for when to allow filtering. Once we decide that it could inform our choice of default.

It seems so universally useful that I'm trying to brainstorm reasons that showing it would be a bad idea.
1. It's a bit more visual noise / cognitive load when interacting with a mouse (have to skip over the filter field, differs from the "typical" select they've used elsewhere on the web)
2. It's possibly more cognitive load when interacting with a keyboard (slightly unintuitive that the arrow keys would change selection when the filter has focus)

For long lists of options those theoretical cons are obviously outweighed by the utility of being able to filter. For short lists maybe it's debatable? It would feel kinda goofy to have a list of 3 items with a filter.

So my initial stab at usage guidance would be "show the filter if there are more than ~6 options". If that's the guidance I'm still leaning towards keeping filtering off by default to reduce churn on existing clients.

But those are all loosely held opinions, I'd be interested in @fredvisser and @atmgrifter00 takes too.

Reply:
I'd agree that `filterMode should default to `none`, as the additional visual complexity is counterproductive for the simple usecases and client apps can easily enable it when appropriate to their data.

- The [USWDS proposes](https://designsystem.digital.gov/components/select/) switching to a filterable option at more than **15 options**
- [NN/g notes](https://www.nngroup.com/articles/listbox-dropdown/) that (non-filterable) select controls should be used when you have between 5-15 options

I'll do a bit more research later, but I can see leveraging that usage guidance for our select as well.

Reply:
I've added a remark that the `none` mode will result in no search input being shown in the dropdown. Whatever usage guidance we come up with will make its way into the Storybook docs with the actual implementation.